Adding the NorthSouth API for Public Preview

MatanShabtay opened the pull request:
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

Commit:
Adding the NorthSouth API for Public Preview

azuresdkci commented:
Can one of the admins verify this patch?

AutorestCI commented:
# Automation for azure-sdk-for-ruby

AutorestCI commented:
# Automation for azure-sdk-for-go

AutorestCI commented:
# Automation for azure-sdk-for-js

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-node

AutorestCI commented:
# Automation for azure-sdk-for-java

Commit:
Fixed some JSON schema validation errors discovered by Open API HUB

Commit:
Fixed bad indenetation + missing API method

Commit:
Added missing NorthSouthResourceName parameter definition

Commit:
Fixed typos found in CR

jhendrixMSFT commented:
@MatanShabtay have these APIs been reviewed by ARM?

MatanShabtay commented:
> @MatanShabtay have these APIs been reviewed by ARM?

This resource type exists in the manifest and it's swagger was reviewed in the PR below, however the PR wasn't completed as this resource is currently in private preview.

We are working towards public preview and following recent changes in the requirements, the API completely changed.
Note that its not considered as a breaking change since we are still under private preview.

https://github.com/Azure/azure-rest-api-specs/pull/3669 

Once you confirm the PR please ask ARM to review as well.
Thanks.

jhendrixMSFT commented:
@ravbhatnagar FYI for ARM review.

Commit:
Added EnforceTrafficHardeningRules action

MatanShabtay commented:
Hi,

Apologies for the late push but I just noticed that I forgot to add a POST action on the resource so I've added it.
Hopefully I didn't cause any trouble.

Thanks in advance.

Commit:
Fixed invalid action name in URI discovered by ARM validation tools (enforce to trafficHardeningsRulesEnforceAction)

KrisBash commented:
@MatanShabtay - let's schedule a quick meeting to understand these changes (ARM api review)

KrisBash commented on specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json:
```diff
@@ -2853,6 +3049,154 @@
+                    "type": "string"
+                },
+                "direction": {
+                    "description": "The rule;s traffic direction",
```
should be an apostrophe rather than semi-colon

KrisBash commented on specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json:
```diff
@@ -2853,6 +3049,154 @@
+                        "$ref": "#/definitions/NorthSouthProtocol"
+                    }
+                },
+                "allowRemoteAddresses": {
```
does 'allow**ed**RemoteAddresses' make more sense than 'allowRemoteAddresses'

MatanShabtay replied:
allowed (past tense) hints that the trafficHardeningRules already applies on the VM's NSG while these set of rules hasn't applied yet on the NSG (some/all of the rules may be enforced on the NSG using the Enforce action).

KrisBash commented on specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json:
```diff
@@ -1438,17 +1438,206 @@
                 }
             }
         },
+        "/subscriptions/{subscriptionId}/providers/Microsoft.Security/northSouthHardenings": {
```
For these GET apis, I would assume that the response could be quite large, as each instance includes a number of rules and a number of alerts. It may be prudent to return only the most relevant data on GET, unless a querystring parameter ($expand) is passed. Additionally, supporting a $filter property (described [here](https://github.com/Microsoft/api-guidelines/blob/vNext/Guidelines.md#97-filtering)) can be helpful to reduce the response size.

KrisBash commented on specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json:
```diff
@@ -1438,17 +1438,206 @@
+            }
+        },
+        "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/{extendedResourceProvider}/{extendedResourceType}/{extendedResourceName}/providers/Microsoft.Security/northSouthHardenings/{northSouthResourceName}": {
+            "get": {
```
How many alerts can there be? Is there any risk of exceeding page size?

KrisBash commented on specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json:
```diff
@@ -1438,17 +1438,206 @@
+                    }
+                }
+            }
+        },
```
This is an async operation, but I don't see the [x-ms-long-running-operation ](https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-long-running-operation) extension

KrisBash commented on specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json:
```diff
@@ -1438,17 +1438,206 @@
+            }
+        },
+        "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/{extendedResourceProvider}/{extendedResourceType}/{extendedResourceName}/providers/Microsoft.Security/northSouthHardenings/{northSouthResourceName}/{trafficHardeningsRulesEnforceAction}": {
+            "post": {
```
I sent a question regarding authorization for this operation offline.

KrisBash reviewed (changes requested):
Please review my comments

jhendrixMSFT commented:
@MatanShabtay any update?

MatanShabtay commented:
Hey,

We have a planned meeting with @KrisBash about the authorization.
We will update the PR and move on after we will make some decisions together.

Commit:
Renamed NorthSouthHardenings to AdaptiveNetworkControls and fixed issues raised in PR

MatanShabtay commented:
1.Due to marketing descisions, the resource type name changed from 'northSouthHardenings' to 'adaptiveNetworkControls'
2.Following further discussions with ARM team - we understood that a resource routingType cannot be both 'Fanout' and 'Extension', so we've decided to stick with extension and removed all subscription and resource group scoped requests.
3.Following our offline discussion with @KrisBash , we have added the NSG IDs to the 'enforce' action request body so we will perform linkedAccessCheck in order to respect RBAC authorizations.
4.Removed the 'alerts' property from the resource - we plan to consume the alerts from a different flow in Microsoft.Security (alerts API)

Commit:
Fixed a bad reference to ListByExtendedResourceAdaptiveNetworkControls_example.json in security.json file

Commit:
Changes AdaptiveNetworkControls protocol property to array instead of string, removed 204 status code from AdaptiveNetworkControls_Enforce

Commit:
Moving 'x-ms-long-running-operation' to the right location

Commit:
Fixing R2005 LongRunningResponseStatusCode error on AdaptiveNetworkControls_Enforce operation

Commit:
Updating EnforceAdaptiveNetworkControlsRules_example.json with the new status code

KrisBash reviewed (approved):
LGTM - thanks!

MatanShabtay commented:
> @MatanShabtay looks good from SDK perspective. Has this functionality been deployed yet?

We are working on updating Microsoft.Security manifest.
Will report back in approx 2weeks (still under development)

jhendrixMSFT commented:
@MatanShabtay any update?  Also there's a merge conflict that needs to be resolved.

MatanShabtay commented:
> @MatanShabtay any update? Also there's a merge conflict that needs to be resolved.

Hi,

Due to marketing decisions the respirce name is about to change from adaptiveNetworkControls to adaptiveNetworkHardenings.

We will soon update the PR with the new resource name.
Should there be any concern about us changing the name, please let me know.

Thanks.

Commit:
Merged from master

Commit:
Renamed AdaptiveNetworkControls to AdaptiveNetworkHardenings due to marketing decisions

Commit:
Merge branch 'Azure-master'

Commit:
Fixing an invalid reference to CloudError

MatanShabtay commented:
Hi @jhendrixMSFT & @KrisBash , Due to marketing decisions I have renamed the resource type name to adaptiveNetworkHardenings from adaptiveNetworkControls & resolved the merge conflicts.

Please review before we will update the manifest and add the new resource type

Commit:
Renaming the parameters and paths so only collection resources will be named in plural

MatanShabtay commented on specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/adaptiveNetworkHardenings.json:
```diff
@@ -0,0 +1,377 @@
+                }
+            }
+        },
+        "AdaptiveNetworkHardenings": {
```
Thanks!
I've renamed the paths and parameters so only collection operation and objects will be named in plural

jhendrixMSFT commented on specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/adaptiveNetworkHardenings.json:
```diff
@@ -0,0 +1,377 @@
+                    "AdaptiveNetworkHardenings"
+                ],
+                "description": "Gets a single Adaptive Network Hardening resource",
+                "operationId": "AdaptiveNetworkHardening_Get",
```
Should be `AdaptiveNetworkHardenings_Get` (missing 's').

MatanShabtay replied:
Why? it gets only a single resource, not a list

jhendrixMSFT replied:
For a given operation ID, AutoRest generates an operation group (e.g. an interface in C#) that's named to everything left of the '_' character that will contain all of the methods in the same group.  So if you have the following operation groups.
```
Foos_Get
Foos_CreateOrUpdate
Foos_Delete
```
You'd get the following codegen.
```C#
public interface Foo {
    Get();
    CreateOrUpdate();
    Delete();
}
```
The other problem here is that operation group name `AdaptiveNetworkHardening` collides with the type `AdaptiveNetworkHardening` defined on line 324; this will force the code generators to disambiguate the name by whatever means the generators have implemented (usually some generic prefix or suffix) which is not very optimal.  It's for this reason that the authoring guidelines state that for operation groups NOUN_VERB the noun be plural so the model type can be singular.

MatanShabtay commented:
Hey @jhendrixMSFT , waiting for your approval.
Can you then refer someone from ARV to review?

Thanks.

Commit:
aligned all operation ids

jhendrixMSFT commented:
@MatanShabtay has this functionality been deployed yet?  I still see DoNotMerge label

Commit:
Adding enforce request body schema

KrisBash reviewed (approved):
Thanks for working with us on this review. LGTM

adxsdknet commented:
# Automation for azure-sdk-for-net 
A PR has been created for you: 
https://github.com/azure/azure-sdk-for-net/pull/5339 
.NET SDK Commits: 
https://github.com/adxsdknet/azure-sdk-for-net/pull/5339/commits/3f8ddd476ad09fbab1ef163efae0a0a36dd6a2bb

jhendrixMSFT commented:
@MatanShabtay hello any update?

MatanShabtay commented:
Hi @jhendrixMSFT ,

The API has been deployed and we ready to complete this PR.
Thanks.

jhendrixMSFT commented:
@MatanShabtay I don't see the new swagger referenced in the readme.md file so some SDK won't get the new APIs generated for them.  Is this intentional?

Commit:
Adding Adaptive Network Hardenings to Security readme.md

MatanShabtay commented:
> @MatanShabtay I don't see the new swagger referenced in the readme.md file so some SDK won't get the new APIs generated for them. Is this intentional?

Hey, I've fixed it.
I see that the Avocado check is failing but based on the error message it doesnt to be related to my PR.

Can you please confirm this ?

jhendrixMSFT commented:
@MatanShabtay it's not but has been fixed in a separate PR.
